Document differences to plain K8s (#159)
appuio/appuio-cloud-docs: docs/modules/ROOT/pages/explanation/kubernetes.adoc

== Key Differences

* <<random-uid,Apps will run with a random UID>>
* <<readonly-fs,Container file systems are read-only>>
* <<network-policies,Containers in different namespaces can't communicate directly by default>>

[#random-uid]
== Random User IDs

[.lead]
Processes in containers started by OpenShift will run with a random user ID for security reasons.

Contrary to your local Docker installation or vanilla Kubernetes clusters, processes in containers started by OpenShift WON'T run with `uid=0` and `gid=0`.
Instead, a random, very high (> 1000000000) UID will be used.

This has some implications, for example:

* Images that try to bind to ports < 1024 won't work
* Images that try to `setuid` during their startup process won't work

Most modern software allows you to configure the ports it binds on, and not take assumptions about user or group IDs.
For example Nginx, in its default configuration, will try to bind to port 80, and then drop to an unprivileged user using `setuid`/`setgid`.
In order to run Nginx in OpenShift, its configuration must be amended to bind to ports > 1024, and any `user` or `group` directives removed.

> Note that if you are using persistent volumes, you don't have to worry about `chown`ing the contents, OpenShift will do this for you automatically when a volume is mounted.

Reviewer: ```suggestion
NOTE: If you are using persistent volumes, you don't have to worry about `chown`ing the contents, OpenShift will do this for you automatically when a volume is mounted.
```

Kubernetes will always make sure that your container processes can write to them.
====

To work around this issue, you can use an https://kubernetes.io/docs/concepts/storage/volumes/#emptydir[`emptyDir` volume].
`emptyDir` will mount an empty directory (what a surprise, given the name) to the path you specified, just as a volume.

=== Example

[source,yaml]
.Basic usage of `emptyDir` volumes
----
apiVersion: apps/v1
kind: Deployment
metadata:
  name: myapp
spec:
  template:
    spec:
      containers:
        - name: myapp
          image: registry.example.com/corp/myapp:v1.5.3
          volumeMounts:
            - name: cache
              mountPath: /var/cache/myapp  # <1>
      volumes:
        - name: cache
          emptyDir:
            sizeLimit: 512Mi  # <2>
----
<1> Mount at whatever path your app needs to write to
<2> `sizeLimit` is optional

[#network-policies]
== Network Namespace Isolation

[.lead]
While on many vanilla Kubernetes clusters, all pods can communicate with each other over the network, OpenShift enforces the use of https://kubernetes.io/docs/concepts/services-networking/network-policies/[Network Policies].

> By default, all INCOMING traffic must be allowed explicitly, while all OUTGOING traffic is allowed.

Reviewer: ```suggestion
On OpenShift, this is only allowed for Pods within the same namespace. Allowing pods from other namespaces to connect, must explicitly be allowed using NetworkPolicies.
```

> While on many vanilla Kubernetes clusters, all pods can communicate with each other over the network, OpenShift enforces the use of https://kubernetes.io/docs/concepts/services-networking/network-policies/[Network Policies].
> By default, all INCOMING traffic must be allowed explicitly, while all OUTGOING traffic is allowed.

Reviewer: ```suggestion
On most Kubernetes clusters, including OpenShift, all pods can communicate with each other through the clusters internal network. APPUiO Cloud by default, only allows this for pods within the same namespace. https://kubernetes.io/docs/concepts/services-networking/network-policies/[Network Policies] must be used to allow incomming trafic from other namespaces. Incomming trafic from ingress is always allowed.
```

[source,yaml]
----
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: allow-from-frontend  # <1>
  namespace: corp-myapp-api-prod  # <2>
spec:
  policyTypes:
    - Ingress
  podSelector:
    matchLabels:
      app: api  # <3>
  ingress:
    - from:
        - namespaceSelector:
            matchLabels:
              kubernetes.io/metadata.name: corp-myapp-frontend-prod  # <4>
          podSelector:
            matchLabels:
              app: frontend  # <5>
      ports:
        - protocol: TCP
          port: 3000  # <6>
----
<1> Give your policy a telling name, so you can identify what it's supposed to do later
<2> The NetworkPolicy must be created in the DESTINATION namespace
<3> Select which DESTINATION pods are affected by this NetworkPolicy
<4> Select which SOURCE namespaces you want to allow traffic from. Labels must be used to select namespaces; the `kubernetes.io/metadata.name` label is added automatically to all namespaces
<5> Select which specific SOURCE pods in the selected namespaces are allowed
<6> Configure which ports traffic is allowed from

> [WARNING]

Reviewer: ```suggestion
[CAUTION]
```

See https://asciidoctor.org/docs/asciidoc-writers-guide/#admonitions.

Author: The guide you linked says

>Use WARNING to inform the reader of **danger, harm, or consequences** that exist.

There is the danger to expose your applications to untrusted third parties, which could cause harm or other consequences, so WARNING seems appropriate?

Author: … but the user must also **act carefully** …

Reviewer: You are probably right. I was more looking at from a "proceed with caution to not mess up" perspective.
